pcap/pcap.go

```diff
@@ -0,0 +1,311 @@
+			p.ui.ErrorLinef("Capture cannot be started on the instance: %s/%s due to error. \nContinue on other instances", host.Job, host.IndexOrID)
+
+			continue
+		}
+
+		runningCaptures++
+
+		packetCs = append(packetCs, packets)
+	}
+
+	if runningCaptures == 0 {
+		return fmt.Errorf("starting of all pcap captures failed")
+	}
+
+	err := writePacketsToFile(opts.SnapLength, opts.Output, packetCs, p.ui)
+	if err != nil {
+		return fmt.Errorf("write to output file failed: %w", err)
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGTERM, syscall.SIGINT, syscall.SIGQUIT)
+
+	select {
+	case <-signals:
+		close(done)
+	case <-ctx.Done():
+		// ctx canceled as cmd exited or an error occurred
+	}
+
+	p.ui.BeginLinef("Wait for session has been finished\n")
+
+	wg.Wait()
+
+	p.ui.EndLinef("\nCapture finished")
+
+	return nil
+}
+
+func writePacketsToFile(snapLength uint32, outputFile string, packetCs []<-chan gopacket.Packet, ui boshui.UI) error {
+	// setup output pcap-file
+	packetFile, err := os.Create(outputFile)
+	if err != nil {
+		return err
+	}
+	packetWriter := pcapgo.NewWriter(packetFile)
+	err = packetWriter.WriteFileHeader(snapLength, layers.LinkTypeEthernet)
+	if err != nil {
+		return err
+	}
+
+	mergedPackets := mergePackets(packetCs)
+	go func() {
+		for packet := range mergedPackets {
+			err = packetWriter.WritePacket(packet.Metadata().CaptureInfo, packet.Data())
+			if err != nil {
+				ui.ErrorLinef("Writing packet to file failed")
+			}
+		}
+		_ = packetFile.Sync()
+		_ = packetFile.Close()
+	}()
+	return nil
+}
+
+func addFilterToCmd(tcpdump, filter, clientIP string, clientSSHPort int) string {
+	filter = strings.TrimSpace(filter)
+	if filter != "" {
+		filter = fmt.Sprintf("not (host %s and port %d) and (%s)", clientIP, clientSSHPort, filter)
+	} else {
+		filter = fmt.Sprintf("not (host %s and port %d)", clientIP, clientSSHPort)
+	}
+	return fmt.Sprintf("%s %q", tcpdump, filter)
+}
+
+func captureSSH(tcpdumpCmd, filter, user string, host boshdir.Host, privateKeyPem ssh.Signer, stopTimeout time.Duration, wg *sync.WaitGroup, done chan struct{}, ui boshui.UI, ctx context.Context, cancel context.CancelCauseFunc) (<-chan gopacket.Packet, error) {
+	client, err := ssh.Dial("tcp", host.Host+":22", &ssh.ClientConfig{
+		Config:          ssh.Config{},
+		User:            user,
+		Auth:            []ssh.AuthMethod{ssh.PublicKeys(privateKeyPem)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         time.Second,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ssh: dial: %w", err)
+	}
+
+	clientSSHAddr, err := getSSHClientIP(client)
+	if err != nil {
+		client.Close()
+		return nil, fmt.Errorf("outbound IP not found %w", err)
+	}
+
+	session, err := client.NewSession()
+	if err != nil {
+		client.Close()
+		return nil, fmt.Errorf("ssh: new session: %w", err)
+	}
+
+	tcpdump := addFilterToCmd(tcpdumpCmd, filter, clientSSHAddr.IP.String(), clientSSHAddr.Port)
+	ui.ErrorLinef(tcpdump)
```

Comment:
I like that it prints the `tcpdump` command that it constructed. Makes it more clear to the user what's being run. 👍